src/python/pants/rules/core/strip_source_roots.py

```diff
@@ -23,26 +32,37 @@ class SourceRootStrippedSources:
 class StripSourceRootsRequest:
   """A request to strip source roots for every file in the snapshot.
 
-  The field `representative_path` is used to determine the source root for the files to be stripped.
-  This assumes that every file shares the same source root, which should be true in practice as the
-  `sources` field for a target always has files in the same source root. We don't proactively
-  validate this assumption because of the performance implications of running
-  `SourceRoots.find_by_path` on every single file in the snapshot, as opposed to only one file. See
-  https://github.com/pantsbuild/pants/pull/9112#discussion_r377999025 for more context on this
-  design.
+  The call site must decide whether to allow the snapshot to include multiple source roots or not,
+  e.g. the snapshot including `src/python/lib.py` and `src/java/lib.java`. If the snapshot
+  is certain to only one have single source root among all the files, then the call site should pass
+  a `representative_path` for better performance. If the snapshot is likely to have more than one
+  snapshot, the call site should set `support_multiple_source_roots=True`, which has worse
+  performance but ensures correctness.
   """
   snapshot: Snapshot
-  representative_path: str
-
-  def determine_source_root(self, *, source_root_config: SourceRootConfig) -> str:
-    source_roots_object = source_root_config.get_source_roots()
-    source_root = source_roots_object.safe_find_by_path(self.representative_path)
-    if source_root is not None:
-      return cast(str, source_root.path)
-    if source_root_config.options.unmatched == "fail":
-      raise NoSourceRootError(f"Could not find a source root for `{self.representative_path}`.")
-    # Otherwise, create a source root by using the parent directory.
-    return PurePath(self.representative_path).parent.as_posix()
+  multiple_source_roots: bool = False
+  representative_path: Optional[str] = None
+
+  def __post_init__(self) -> None:
+    if self.multiple_source_roots and self.representative_path is not None:
```

Comment:
Instead of this complication, can we have `representative_path=None` imply that there can be multiple source roots, and get rid of `multiple_source_roots` entirely?

Reply:
Even simpler, can this all just always strip all source roots given only a Snapshot? Fwict the 1 caller has that snapshot and need not pass representative_path:
https://github.com/pantsbuild/pants/blob/d0998872d8116edd5ebc8231dbdd28779591dc2b/src/python/pants/rules/core/find_target_source_files.py#L45-L53

Reply:
Always stripping was the original behavior and we decided to not do it for a performance cost. It’s non-trivial to run SourceRoots.find_by_path on every file. 

Benjy, that’s fine with me to remove the Boolean, but I’d want to make rep_path no longer be an optional field so that you must make an explicit decision.

Reply:
Yes indeed- it would no longer be optional, and it would need to be documented that the representative is a performance hack in the case where you know there is only one source root.

Reply:
Well, it would still be optional in that it can be a string or None. Unless you want to use an empty string as the sentinel value?

Reply:
@benjyw I implemented the suggestion. Can you take another look, please?

I ended up giving it a default value of `None` with the idea that the rule should support multiple source roots by default. That is the sane default behavior. We only provide a performance hack that you must expliclity opt in to. 

It looked weird to read in a call site `StripSourceRootsRequest(snapshot, representative_path=None)`. What does `representative_path=None` mean..? It's not very clear that it implied "There is no one single representative path".